src/options/dir_action.rs

```diff
@@ -54,79 +54,496 @@ impl RecurseOptions {
+            paths: vec![],
+            all: 0,
+            long: 0,
+            git: 0,
+            oneline: 0,
+            recurse: 0,
+            list_dirs: 0,
+            tree: 0,
+            level: None,
+            reverse: 0,
+            sort: None,
+            ignore_glob: None,
+            git_ignore: 0,
+            dirs_first: 0,
+            only_dirs: 0,
+            binary: 0,
+            bytes: 0,
+            group: 0,
+            numeric: 0,
+            grid: 0,
+            across: 0,
+            classify: 0,
+            dereference: 0,
+            width: None,
+            color: None,
+            color_scale: 0,
+            almost_all: 0,
+            header: 0,
+            icons: 0,
+            inode: 0,
+            git_repos: 0,
+            git_repos_no_status: 0,
+            links: 0,
+            modified: 0,
+            created: 0,
+            accessed: 0,
+            changed: 0,
+            blocksize: 0,
+            time: None,
+            time_style: None,
+            no_filesize: 0,
+            no_icons: 0,
+            no_permissions: 0,
+            no_time: 0,
+            no_user: 0,
+            extended: 0,
+            hyperlink: 0,
+            octal: 0,
+            security_context: 0,
+            help: false,
+        };
+
+        assert_eq!(DirAction::deduce(&matches, false, false).unwrap(), DirAction::List);
+    }
+
+    #[test]
+    fn deduce_recurse() {
+        let matches = Opts {
+            paths: vec![],
+            all: 0,
+            long: 0,
+            git: 0,
+            oneline: 0,
+            recurse: 1,
+            list_dirs: 0,
+            tree: 0,
+            level: None,
+            reverse: 0,
+            sort: None,
+            ignore_glob: None,
+            git_ignore: 0,
+            dirs_first: 0,
+            only_dirs: 0,
+            binary: 0,
+            bytes: 0,
+            group: 0,
+            numeric: 0,
+            grid: 0,
+            across: 0,
+            classify: 0,
+            dereference: 0,
+            width: None,
+            color: None,
+            color_scale: 0,
+            almost_all: 0,
+            header: 0,
+            icons: 0,
+            inode: 0,
+            git_repos: 0,
+            git_repos_no_status: 0,
+            links: 0,
+            modified: 0,
+            created: 0,
+            accessed: 0,
+            changed: 0,
+            blocksize: 0,
+            time: None,
+            time_style: None,
+            no_filesize: 0,
+            no_icons: 0,
+            no_permissions: 0,
+            no_time: 0,
+            no_user: 0,
+            extended: 0,
+            hyperlink: 0,
+            octal: 0,
+            security_context: 0,
+            help: false,
+        };
+        assert_eq!(DirAction::deduce(&matches, false, false).unwrap(), DirAction::Recurse(RecurseOptions {
+            tree: false,
+            max_depth: None,
+        }));
+    }
```

Comment:
You should be able to do avoid a lot of the code duplication with something like:

```suggestion
    let default_opts = {
        paths: vec![],
        all: 0,
        long: 0,
        git: 0,
        oneline: 0,
        recurse: 0,
        list_dirs: 0,
        tree: 0,
        level: None,
        reverse: 0,
        sort: None,
        ignore_glob: None,
        git_ignore: 0,
        dirs_first: 0,
        only_dirs: 0,
        binary: 0,
        bytes: 0,
        group: 0,
        numeric: 0,
        grid: 0,
        across: 0,
        classify: 0,
        dereference: 0,
        width: None,
        color: None,
        color_scale: 0,
        almost_all: 0,
        header: 0,
        icons: 0,
        inode: 0,
        git_repos: 0,
        git_repos_no_status: 0,
        links: 0,
        modified: 0,
        created: 0,
        accessed: 0,
        changed: 0,
        blocksize: 0,
        time: None,
        time_style: None,
        no_filesize: 0,
        no_icons: 0,
        no_permissions: 0,
        no_time: 0,
        no_user: 0,
        extended: 0,
        hyperlink: 0,
        octal: 0,
        security_context: 0,
        help: false,
    };

    #[test]
    fn deduces_list() {
        let matches = default_opts;
    }

    #[test]
    fn deduce_recurse() {
        let matches = Opts {
            recurse: 1,
            ..default_opts,
        };
        assert_eq!(DirAction::deduce(&matches, false, false).unwrap(), DirAction::Recurse(RecurseOptions {
            tree: false,
            max_depth: None,
        }));
    }
```

btw I like that it get rids of some test with macros which aren’t very readable or easy to maintain in my opinion.

Reply:
Yeah i didnt like the macros too
Oh didnt tought of that will do it